File: cpp/structures/LinkedList.cpp
#include <iostream>
#include <ostream>

class Node {
 public:
  int value;
  Node* next;

  Node(int value) {
    this->value = value;
    this->next = nullptr;
  }
};

class LinkedList {
 private:
  Node* head;
  Node* tail;
  int length;

 public:
  LinkedList(int value) {
    Node* newNode = new Node(value);
    this->head = newNode;
    this->tail = newNode;
    this->length = 1;
  }

  Node* get(int index) {
    if (index < 0 || index >= this->length) return nullptr;

    Node* temp = this->head;

    for (int i = 0; i < index; i++) {
      temp = temp->next;
    }

    return temp;
  }

  bool set(int index, int value) {
    Node* temp = this->get(index);

    if (temp != nullptr) {
      temp->value = value;
      return true;
    } else {
      return false;
    }
  }

  void append(int value) {
    Node* newNode = new Node(value);

    if (this->length == 0) {
      this->head = newNode;
      this->tail = newNode;
    } else {
      this->tail->next = newNode;
      this->tail = newNode;
    }

    this->length += 1;
  }

  void prepend(int value) {
    Node* newNode = new Node(value);

    if (this->length == 0) {
      this->head = newNode;
      this->tail = newNode;
    } else {
      newNode->next = this->head;
      this->head = newNode;
    }

    this->length += 1;
  }

  bool insert(int index, int value) {
    if (index < 0 || index > this->length) return false;

    if (index == 0) {
      this->prepend(value);
      return true;
    }

    if (index == this->length) {
      this->append(value);
      return true;
    }

    Node* newNode = new Node(value);
    Node* temp = this->get(index - 1);

    newNode->next = temp->next;
    temp->next = newNode;

    this->length += 1;

    return true;
  }

  Node* remove(int index) {
    if (index < 0 || index >= this->length) return nullptr;

    if (index == 0) {
      return this->removeFirst();
    }

    if (index == this->length - 1) {
      return this->removeLast();
    }

    Node* prev = this->get(index - 1);
    Node* temp = prev->next;

    prev->next = temp->next;
    temp->next = nullptr;

    this->length -= 1;

    return temp;
  }

  Node* removeFirst() {
    if (this->length == 0) return nullptr;

    Node* temp = this->head;
    this->head = this->head->next;
    temp->next = nullptr;

    this->length -= 1;

    if (this->length == 0) {
      this->tail = nullptr;
    }

    return temp;
  }

  Node* removeLast() {
    Node* temp = this->head;
    Node* prev = this->head;

    if (this->length == 0) return nullptr;

    while (temp->next != nullptr) {
      prev = temp;
      temp = temp->next;
    }

    this->tail = prev;
    this->tail->next = nullptr;
    this->length -= 1;

    if (this->length == 0) {
      this->head = nullptr;
      this->tail = nullptr;
    }

    return temp;
  }

  void reverse() {
    Node* temp = this->head;
    this->head = this->tail;
    this->tail = temp;

    Node* after = temp->next;
    Node* before = nullptr;

    for (int i = 0; i < this->length; i++) {
      after = temp->next;
      temp->next = before;
      before = temp;
      temp = after;
    }
  }

  void print() {
    Node* temp = this->head;
    while (temp != nullptr) {
      std::cout << temp->value << " -> ";
      temp = temp->next;
    }

    std::cout << "nullptr" << std::endl;
  }

  void getHead() { std::cout << "Head: " << this->head->value << std::endl; }
  void getTail() { std::cout << "Tail: " << this->tail->value << std::endl; }
  void getLength() { std::cout << "Length: " << this->length << std::endl; }
};

int main(int argc, char* argv[]) {
  LinkedList* list = new LinkedList(0);
  list->append(1);
  list->append(2);
  list->append(3);
  list->append(4);
  list->print();

  list->reverse();
  list->print();

  delete list;

  return 0;
}

// clang++ -std=c++20 LinkedList.cpp -o LinkedList ; ./LinkedList

File: python/structures/linked_list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class LinkedList:
    def __init__(self, value):
        node = Node(value)
        self.head = node
        self.tail = node
        self.length = 1

    def append(self, value) -> bool:
        node = Node(value)
        if self.length == 0:
            self.head = node
            self.tail = node
        else:
            self.tail.next = node
            self.tail = node
        self.length += 1

        return True

    def pop(self) -> Node | None:
        if self.length == 0:
            return None

        temp = self.head
        prev = self.head

        while temp.next:
            prev = temp
            temp = temp.next

        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            self.tail = prev
            self.tail.next = None
        self.length -= 1

        return temp

    def prepend(self, value) -> bool:
        node = Node(value)
        if self.length == 0:
            self.head = node
            self.tail = node
        else:
            node.next = self.head
            self.head = node
        self.length += 1

        return True

    def pop_first(self) -> Node | None:
        if self.length == 0:
            return None

        temp = self.head

        if self.length == 1:
            self.tail = None
        else:
            self.head = self.head.next
            temp.next = None
        self.length -= 1

        return temp

    def get(self, index) -> Node | None:
        if index < 0 or index >= self.length:
            return None

        temp = self.head

        for _ in range(index):
            temp = temp.next

        return temp

    def set_value(self, index, value) -> bool | None:
        temp = self.get(index)

        if temp is not None:
            temp.value = value

        return True

    def insert(self, index, value) -> bool:
        if index < 0 or index > self.length:
            return False

        if index == 0:
            return self.prepend(value)

        if index == self.length:
            return self.append(value)

        node = Node(value)
        temp = self.get(index - 1)
        node.next = temp.next
        temp.next = node

        return True

    def remove(self, index) -> Node | None:
        if index < 0 or index >= self.length:
            return None

        if index == 0:
            return self.pop_first()

        if index == self.length - 1:
            return self.pop()

        prev = self.get(index - 1)
        temp = prev.next
        prev.next = temp.next
        temp.next = None
        self.length -= 1

        return temp

    def reverse(self):
        temp = self.head
        self.head = self.tail
        self.tail = temp

        after = temp.next
        before = None

        for _ in range(self.length):
            after = temp.next
            temp.next = before
            before = temp
            temp = after

    def print(self):
        temp = self.head
        while temp is not None:
            print(temp.value)
            temp = temp.next


if __name__ == "__main__":
    list = LinkedList(0)
    list.append(1)
    list.append(2)
    list.append(3)
    list.insert(4, 4)
    list.reverse()

    list.print()
